Manually adding an external related article

MaciekBaron opened the pull request:
### What is the context of this PR?

This PR addresses CMS-461 and allows the user to add external links to related articles in a topic page.

**Loom video**
https://www.loom.com/share/70c1009a0f07439d9e67760b5e9828a1?sid=3569f082-2a60-491f-85ad-aa3504c56b84

### How to review

1. Make sure that you have a topic page which has at least three different article series, each with at least one article
2. Go to the topic page, ensure that the related articles appear
3. Edit the topic page and add an external link to the related articles
4. Look at the topic page, ensure that the manually added link appears first, followed by 2 internal related pages
5. Add another external link
6. Look at the topic page, ensure that there are 2 manually added links and one internal page
7. Add another external link
8. Look at the topic page, ensure that all 3 manually added links appear and no internal pages are listed
9. Remove all external links
10. Look at the topic page, ensure that 3 internal pages are now listed

### Follow-up Actions

- Check if a description field is required to ensure consistency with internal pages
- Check if "Article" is an appropriate type for external links

sanjeevz3009 commented:
<img width="1016" height="559" alt="Screenshot 2025-07-30 at 11 54 03" src="https://github.com/user-attachments/assets/130f4f5d-4bf0-498b-b026-f4a53edc1826" />

@MaciekBaron This looks odd.

Why do we want?
- `Internal link` and then `Page`
- `External Link` and then `External URL`

One or another should do.

I particularly would go with `Page` and `External URL`.

sanjeevz3009 commented:
Did the initial round of reviews:

@MaciekBaron All the core functionalities are working well, good test coverage, both unit and UI tests.

When I `Add topic page related article`:
- I select an internal page
- I am able to enter a title even though I have selected an internal page. It doesn't throw a validation error, as I would have expected, as adding a title here doesn't seem to do anything or override the internal page title anyway, so this is misleading

I know we discussed this over on Slack:
- The JIRA card requirement says, `Then the manually added link(s) should be displayed above the auto populated links`
- This is not the case, as when we add an `internal page` first, then that will show above the added `external link`. Some clarification would be nice here.

MaciekBaron commented on cms/topics/models.py:
```diff
@@ -159,46 +194,60 @@ def latest_article_in_featured_series(self) -> StatisticalArticlePage | None:
+        manual_articles = []
+        highlighted_page_pks = []
+
+        for related in self.related_articles.select_related("page").all():
```
Restructured in [b29e8fa](https://github.com/ONSdigital/dis-wagtail/pull/331/commits/b29e8fa0c88fa92efdd5351888e6e11c8010f5db)

MaciekBaron commented:
Re. https://github.com/ONSdigital/dis-wagtail/pull/331#issuecomment-3135802969:

I've changed the wording from `External URL` to just `URL` so that it reads better. So now it's Internal link -> Page and External Link -> URL which I think makes more sense. Happy to discuss further.

---
Re. https://github.com/ONSdigital/dis-wagtail/pull/331#issuecomment-3135881825:

Regarding the first point, that's a good shout and I have added additional validation in [b29e8fa](https://github.com/ONSdigital/dis-wagtail/pull/331/commits/b29e8fa0c88fa92efdd5351888e6e11c8010f5db)

The second point I think we might need to clarify with the team. From what I understood, manually added links is a set which includes internal pages and external links (both manually added). The AC doesn't distinguish between external and internal links, so my assumption is that they are one and the same here. But, like I said, worth checking as my interpretation might be false.

MaciekBaron commented on cms/topics/models.py:
```diff
@@ -43,14 +43,52 @@ class TopicPageRelatedArticle(Orderable):
+        if self.page:
+            if self.external_url:
+                raise ValidationError("Please select either an internal page or provide an external URL, not both.")
+            if self.title:
```
I had to refactor the code slightly to allow passing the title further down here: [5104371](https://github.com/ONSdigital/dis-wagtail/pull/331/commits/510437177731cc81b3e300d4bb1dbd1c51ba22ae)

AdamHawtin reviewed (approved):
LGTM 👍

sanjeevz3009 reviewed (commented):
Re-reviewed! All looks good to me. Thank you for making the requested changes.

I did find some other issues that are out of the scope of this PR on the topic page.
- You can select duplicate methodology pages under the `Highlighted methods and quality information`, when adding `Topic page related methodologies`.
  - I don't think we should be allowing this + it doesn't even show the duplicate methodology pages.
    - If we have selected 3 methodology pages, and there are 2 duplicates -> the website shows 3 unique methodology pages, regardless of including one from the duplicates as it replaces one of the duplicates with another methodology page if it exists
    - We just need to make sure in Wagtail we don't allow this behaviour, have validations against this, as this is clearly misleading
    
Other than that, I also found out we seem to be auto-populating to show the 3 oldest methodology pages, so the first 3 methodology pages created, for example, and the 4th one won't autopopulate. Not sure if this is the correct behaviour, as I would expect the 3 latest methodology pages to show.

sanjeevz3009 reviewed (approved):
Other than the conversations + comments added (out of scope). Happy to approve =)
